File: backend/src/models/Review.py
from __future__ import annotations

from typing import Optional

from beanie import Indexed
from beanie.odm.fields import PydanticObjectId
from pydantic import Field, model_validator

from .base import TimeStampedDocument


class Review(TimeStampedDocument):
    # Support both naming conventions
    user: Indexed(PydanticObjectId)  # type: ignore[assignment]
    userId: Optional[PydanticObjectId] = None
    
    product: Indexed(PydanticObjectId)  # type: ignore[assignment]
    productId: Optional[PydanticObjectId] = None
    
    rating: int = Field(ge=1, le=5)  # 1-5 stars
    title: Optional[str] = Field(default=None, max_length=100)
    
    # Support both naming conventions for review text
    comment: Optional[str] = Field(default=None, max_length=1000)
    reviewText: Optional[str] = Field(default=None, max_length=1000)
    
    isVerifiedPurchase: bool = False
    isApproved: bool = Field(default=True, index=True)  # Admin can moderate reviews
    isDeleted: bool = False
    helpfulCount: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def sync_fields(self) -> "Review":
        # Sync user fields
        if self.userId is None:
            self.userId = self.user
        
        # Sync product fields
        if self.productId is None:
            self.productId = self.product
        
        # Sync comment and reviewText
        if self.comment and not self.reviewText:
            self.reviewText = self.comment
        elif self.reviewText and not self.comment:
            self.comment = self.reviewText
            
        return self

    class Settings:
        name = "reviews"
        use_revision = False
        indexes = [
            [("product", 1), ("user", 1)],  # Unique review per user per product
            [("productId", 1), ("userId", 1)],  # Alternative naming
            [("product", 1), ("createdAt", -1)],  # Sort by recent for product
            [("user", 1), ("createdAt", -1)],  # Sort by recent for user
            [("isApproved", 1)],  # Filter by approval status
        ]

File: admin-backend/src/models/Review.js
import mongoose from 'mongoose';

const reviewSchema = new mongoose.Schema({
  // Support both naming conventions
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  productId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Product',
    required: true,
    index: true
  },
  product: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Product'
  },
  rating: {
    type: Number,
    required: true,
    min: 1,
    max: 5
  },
  title: {
    type: String,
    trim: true,
    maxlength: 100
  },
  reviewText: {
    type: String,
    trim: true,
    maxlength: 1000
  },
  comment: {
    type: String,
    trim: true,
    maxlength: 1000
  },
  isVerifiedPurchase: {
    type: Boolean,
    default: false
  },
  isApproved: {
    type: Boolean,
    default: true,
    index: true
  },
  isDeleted: {
    type: Boolean,
    default: false
  },
  helpfulCount: {
    type: Number,
    default: 0,
    min: 0
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Create compound index for unique review per user per product
reviewSchema.index({ productId: 1, userId: 1 });
reviewSchema.index({ product: 1, user: 1 });

// Sync field names for compatibility
reviewSchema.pre('save', function(next) {
  // Sync user fields
  if (this.userId && !this.user) {
    this.user = this.userId;
  } else if (this.user && !this.userId) {
    this.userId = this.user;
  }
  
  // Sync product fields
  if (this.productId && !this.product) {
    this.product = this.productId;
  } else if (this.product && !this.productId) {
    this.productId = this.product;
  }
  
  // Sync comment and reviewText
  if (this.reviewText && !this.comment) {
    this.comment = this.reviewText;
  } else if (this.comment && !this.reviewText) {
    this.reviewText = this.comment;
  }
  
  this.updatedAt = Date.now();
  next();
});

const Review = mongoose.model('Review', reviewSchema, 'reviews');

export default Review;
